File: domain/src/main/java/by/grodno/vasili/domain/interactor/UseCase.java
package by.grodno.vasili.domain.interactor;

import io.reactivex.disposables.CompositeDisposable;


/**
 * Abstract class for a Use Case.
 */
abstract class UseCase<T, Param> {
    /**
     * {@link CompositeDisposable} for cancel pending executions
     */
    abstract CompositeDisposable getDisposables();

    /**
     * Execute use case with params and observer
     *
     * @param observer for observing result of execution use case
     * @param params   for executing use case
     */
    abstract void execute(T observer, Param params);

    /**
     * Dispose execution
     */
    public void dispose() {
        CompositeDisposable disposables = getDisposables();
        if (disposables != null && !disposables.isDisposed()) {
            disposables.dispose();
        }
    }
}

File: domain/src/main/java/by/grodno/vasili/domain/interactor/UseCase.kt
package by.grodno.vasili.domain.interactor

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.cancel

/**
 * Base use case for application needs. Execute some action with [Params] and return the result with [Type].
 *
 */
abstract class UseCase<Params, Type> : CoroutineScope by CoroutineScope(Dispatchers.IO) {

    /**
     * Action for retrieving data.
     */
    protected abstract suspend fun action(params: Params): Type

    /**
     * Run action.
     */
    suspend fun execute(params: Params): Type = action(params)

    /**
     * Cancel use case execution.
     */
    fun dispose() {
        cancel()
    }
}
